### PR TITLE
Add authentication

### LINKED ISSUE #29: Implement google authentication

### PR DESCRIPTION
**Pull request by andres-vidal**
Closes #29 

Refer to the [branch's README](https://github.com/wyeworks/richardburton/tree/add-authentication) for the specification of the implemented flow.

### CODE REVIEW

File: backend/lib/richard_burton/auth_google.ex [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      {%{"issuer" => issuer}, keys} ->
+        verify(
+          token,
+          issuer: issuer,
+          audience: get_audience(),
+          keys: keys
+        )
+
+      _ ->
+        throw("Auth configuration is not properly set.s")
```
Comment (iaguirre88) [inline backend/lib/richard_burton/auth_google.ex:26]:
```suggestion
        throw("Auth configuration is not properly set.")
```

File: backend/lib/richard_burton/auth_google.ex [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  @impl true
+  @spec init() :: {Map.t(), List.t()}
+  def init(), do: {get_config(), get_keys()}
+
+  @impl true
+  @spec verify(token :: String.t()) :: {:ok, String.t()} | :error
+  def verify(token) do
+    case Application.get_env(:richard_burton, :auth_config) do
+      {%{"issuer" => issuer}, keys} ->
+        verify(
```
Comment (iaguirre88) [inline backend/lib/richard_burton/auth_google.ex:18]: NIT: even though `verify/1` and `verify/4` are different functions, I think it's more idiomatic to name the private one something like `do_verify` so it's clearer for the reader which one is the public one.

File: backend/test/richard_burton/user_test.exs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    test "when subject id is blank, is invalid" do
+      refute change_valid(%{"subject_id" => ""}).valid?
+    end
+
+    test "when subject id is nil is invalid" do
+      refute change_valid(%{"subject_id" => nil}).valid?
+    end
+
+    test "when an author with the provided attributes exist, is invalid" do
```
Comment (iaguirre88) [inline backend/test/richard_burton/user_test.exs:48]: > when a **user** with ...

File: backend/test/richard_burton_web/controllers/user_controller_test.exs [hunk 4]
```diff
@@ -0,0 +1,42 @@
+defmodule RichardBurtonWeb.UserControllerTest do
+  @moduledoc """
+  Tests for the Publication controller
+  """
+  use RichardBurtonWeb.ConnCase
+  import Routes, only: [user_path: 2]
+
+  @valid_attrs %{"email" => "example@gmail.com"}
+  @valid_return %{"email" => "example@gmail.com", "role" => "reader"}
```
Comment (iaguirre88) [inline backend/test/richard_burton_web/controllers/user_controller_test.exs:9]: NIT: maybe `@expected_return`?

File: backend/test/richard_burton_web/controllers/user_controller_test.exs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+  describe "POST /users" do
+    test "when providing valid params, returns 201 created and the inserted user", meta do
```
Comment (iaguirre88) [inline backend/test/richard_burton_web/controllers/user_controller_test.exs:12]: NIT:
```suggestion
    test "when providing valid params, returns 201 created and the inserted user", %{conn: conn} do
```

**Review by iaguirre88 — COMMENTED** [review]
I just left a first round of comments.

I've noticed some changes related to `Author`, `OriginalBook`, `Publication`, and `TranslatedBook`. Were those changes introduced here? Or is it because the branch needs rebasing?

Let me know so I can decide whether I should review them or not.

**Comment by andres-vidal** [comment]
@iaguirre88 indeed, the branch needs rebasing 😅

**Comment by andres-vidal** [comment]
@iaguirre88 there it goes, addressed all your comments. LMK what you think, specially about the rewrite on the README. Thank's for carefully reading it 😊

**Pull request merged**